modules/stanford_events_importer/src/Plugin/Field/FieldWidget/StanfordEventsImporterAPIURLFieldWidget.php

```diff
@@ -0,0 +1,238 @@
+    ];
+
+    // The organization id (integer) as provided in the xml feed.
+    $element['_other']['organization'] = [
+      '#type' => 'select',
+      '#title' => $this->t("Organization"),
+      '#field_parents' => ['other'],
+      '#options' => $this->getOrgOptions(),
+      '#states' => [
+        'visible' => [
+          '#edit-su-event-xml-url-' . $delta . '-other-type' => ['value' => 'organization'],
+        ],
+      ],
+      '#default_value' => $defaults['organization'] ?? '',
+    ];
+
+    $element['_other']['category'] = [
+      '#type' => 'select',
+      '#title' => $this->t("Category"),
+      '#field_parents' => ['other'],
+      '#options' => $this->getCatOptions(),
+      '#states' => [
+        'visible' => [
+          '#edit-su-event-xml-url-' . $delta . '-other-type' => ['value' => 'category'],
+        ],
+      ],
+      '#default_value' => $defaults['category'] ?? '',
+    ];
+
+    $element['_other']['org_status'] = [
+      '#type' => 'select',
+      '#title' => $this->t("Event Status"),
+      '#field_parents' => ['other'],
+      '#options' => [
+        'published' => $this->t("Published"),
+        'unlisted' => $this->t("Unlisted"),
+        'bookmarked' => $this->t("Bookmarked"),
+      ],
+      '#states' => [
+        'visible' => [
+          '#edit-su-event-xml-url-' . $delta . '-other-type' => ['value' => 'organization'],
+        ],
+      ],
+      '#default_value' => $defaults['org_status'] ?? '',
+    ];
+
+    // Hide the uri.
+    $element['uri']['#type'] = "hidden";
+    $element['uri']['#required'] = FALSE;
+    return $element;
+  }
+
+  /**
+   * Get a key/value list of organizations from the API.
+   *
+   * @return array
+   *   An array of select form options.
+   */
+  protected function getOrgOptions() {
+    $opts = \Drupal::state()->get("stanford_events_importer_orgs");
+
+    // If state doesn't exist. Try to get it.
+    if (empty($opts)) {
+      stanford_events_importer_update_opts();
+      $opts = \Drupal::state()->get("stanford_events_importer_orgs");
+    }
+
+    array_unshift($opts, $this->t("- Select Organization -"));
+    return $opts;
+  }
+
+  /**
+   * Get a key/value list of categories from the API.
+   *
+   * @return array
+   *   An array of select form options.
+   */
+  protected function getCatOptions() {
+
+    $opts = \Drupal::state()->get("stanford_events_importer_cats");
+
+    // If state doesn't exist. Try to get it.
+    if (empty($opts)) {
+      stanford_events_importer_update_opts();
+      $opts = \Drupal::state()->get("stanford_events_importer_cats");
+    }
+
+    array_unshift($opts, $this->t("- Select Category -"));
+    return $opts;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function massageFormValues(array $values, array $form, FormStateInterface $form_state) {
+
+    // Parse form options we added into a url for the events.stanford.edu feed.
+    array_walk($values, function (&$value) {
```

Comment:
thats a big array walk function. i'd suggest moving that into its own function in this class. you can even do a `protected static function` and array walk will still run it.